src/state-manager/TransactionConsensus.ts

```diff
@@ -1135,12 +1147,38 @@ class TransactionConsenus {
             '',
             true
           )
+
           queueEntry.hasSentFinalReceipt = true
+
+          // If the queue entry does not have the valid final data then request that
+          if (!queueEntry.hasValidFinalData) {
+            setTimeout(async () => {
+              // Check if we have final data
+              if (queueEntry.hasValidFinalData) {
+                return
+              }
+              if (logFlags.verbose)
+                this.mainLogger.debug(`poqo-receipt-gossip: requesting final data for ${queueEntry.logID}`)
+              nestedCountersInstance.countEvent(
+                'request-final-data',
+                'final data timeout, making explicit request'
+              )
+
+              const nodesToAskKeys = payload.signatures?.map((signature) => signature.owner)
+
+              await this.stateManager.transactionQueue.requestFinalData(
```

Comment:
requestFinalData can return based on number of accounts found in collectedFinalData === payload.appliedVote.account_id.length which is different from queueEntry.hasValidFinalData. Potentially can be infinite loop.

Reply:
I have enhanced out counters and added some to our doc for sys-automation monitoring.

I have created a blue task to resolve this.  the resolution will be that we need to check state information before putting something in the collected final data map.  there is no value in putting a wrong state account this loop which can cause logic errors like the one you found above.

Created blue-203 to resolve